MapboxNavigation/InstructionPresenter.swift

```diff
@@ -83,37 +83,44 @@ class InstructionPresenter {
                 strings.append(attributedStrings.reduce(initial, +))
             }
             
+            switch component.type {
             //Throw away exit components. We know this is safe because we know that if there is an exit component,
             //  there is an exit code component, and the latter contains the information we care about.
-            guard component.type != .exit else { continue }
-            
-            //If we have a exit, in the first two components, lets handle that first.
-            if component.type == .exitCode, 0...1 ~= index,
-                let exitString = attributedString(forExitComponent: component, maneuverDirection: instruction.maneuverDirection, dataSource: dataSource) {
+            case .exit:
+                continue
+                
+            //If we have a exit, in the first two components, lets handle that.
+            case .exitCode where 0...1 ~= index:
+                guard let exitString = self.attributedString(forExitComponent: component, maneuverDirection: instruction.maneuverDirection, dataSource: dataSource) else { fallthrough }
                 build(component, [exitString])
-            }
                 
-            //If we have a shield, lets include those
-            else if let shieldString = attributedString(forShieldComponent: component, repository: imageRepository, dataSource: dataSource, onImageDownload: onImageDownload) {
-                build(component, [joinString, shieldString])
-            }
-            
-            else {
-                //if it's a delimiter, skip it if it's between two shields. Otherwise, process the regular text component.
-                if component.type == .delimiter {
-                    
-                    let componentBefore = components.component(before: component)
-                    let componentAfter = components.component(after: component)
-                    
-                    if let shieldKey = componentBefore?.cacheKey(),
-                        imageRepository.cachedImageForKey(shieldKey) != nil {
-                        continue
-                    }
-                    if let shieldKey = componentAfter?.cacheKey(),
-                        imageRepository.cachedImageForKey(shieldKey) != nil {
-                        continue
-                    }
+            //If we have an icon component, lets turn it into a shield.
+            case .icon:
+                if let shieldString = attributedString(forShieldComponent: component, repository: imageRepository, dataSource: dataSource, onImageDownload: onImageDownload) {
+                    build(component, [joinString, shieldString])
+                } else if let genericShieldString = attributedString(forGenericShield: component, dataSource: dataSource) {
+                    build(component, [joinString, genericShieldString])
+                } else {
+                    fallthrough
```

Comment:
This is a clever use of a switch statement, but do you think the fallthroughs inside conditionals make the code flow a little more difficult to keep track of?

Reply:
@1ec5 It's certainly a complex switch, but I'd make the case that it's a heck of a lot easier to follow than the mess of `if` statements before it.